### PR TITLE
Added service discovery

### PR DESCRIPTION
**Pull request by mshariq-nerd**
Dependencies to run this PR:

1. service-discovery server must be up 
2. service name must be unique

### CODE REVIEW

**Comment by praveen-garg** [comment]
Move it separate menu/ route/ component

File: src/app/service-discovery/service-discovery.component.ts [hunk 1]
```diff
@@ -29,24 +38,71 @@ export class ServiceDiscoveryComponent implements OnInit, AfterViewInit {
...
           serviceRecord.push({
             key: key,
             [key]: data[key]
           })
         });
         this.serviceData = serviceRecord;
+        if (!this.serviceStatus && this.serviceData.length != 0) {
+          this.message = 'Connected service is down. Connect to other service listed below or you can connect a service manually from <a href="/setting">settings</a>.'
+        } else {
+          this.message = 'Connected service is down. You can connect a service manually from <a href="/setting">settings</a>.'
```
Comment (praveen-garg) [inline src/app/service-discovery/service-discovery.component.ts]: I assume there will be 3 states

File: src/app/service-discovery/service-discovery.component.ts [hunk 2]
```diff
@@ -29,24 +38,71 @@ export class ServiceDiscoveryComponent implements OnInit, AfterViewInit {
...
       (data) => {
+        this.serviceDiscoveryStatus = true;
         Object.keys(data).forEach(function (key) {
           serviceRecord.push({
             key: key,
             [key]: data[key]
           })
         });
         this.serviceData = serviceRecord;
+        if (!this.serviceStatus && this.serviceData.length != 0) {
```
Comment (praveen-garg) [inline src/app/service-discovery/service-discovery.component.ts]: serviceStatus > connectedServiceStatus
serviceData > discoveredServices or discoveredServicesData

File: src/app/service-discovery/service-discovery.component.ts [hunk 3]
```diff
@@ -29,24 +38,71 @@ export class ServiceDiscoveryComponent implements OnInit, AfterViewInit {
...
+        } else {
+          this.message = 'Connected service is down. You can connect a service manually from <a href="/setting">settings</a>.'
+        }
       },
       (error) => {
-        console.log('error: ', error);
+        this.serviceDiscoveryStatus = false;
+        this.serviceData = [];
+        if (error.status === 0) {
+          this.message = 'Not able to connect. Please check service discovery server up and running.'
```
Comment (praveen-garg) [inline src/app/service-discovery/service-discovery.component.ts]: ... is up and running.

File: src/app/service-discovery/service-discovery.component.ts [hunk 4]
```diff
@@ -29,24 +38,71 @@ export class ServiceDiscoveryComponent implements OnInit, AfterViewInit {
...
+          console.log('service down ', error);
+        } else {
+          this.message = "Something wrong at service discovery side. Try again later."
```
Comment (praveen-garg) [inline src/app/service-discovery/service-discovery.component.ts]: Something wrong with the discovery service. Try again!

File: src/app/service-discovery/service-discovery.component.ts [hunk 5]
```diff
@@ -29,24 +38,71 @@ export class ServiceDiscoveryComponent implements OnInit, AfterViewInit {
...
+          console.log('error in response ', error);
+        }
       });
   }
 
+  start() {
+    clearInterval(this.timer);
+    this.timer = setInterval(function () {
+      this.discoverService();
+    }.bind(this), POLLING_INTERVAL);
```
Comment (praveen-garg) [inline src/app/service-discovery/service-discovery.component.ts]: we should use DISCOVERY_SERVICE_POLLING_INTERVAL

File: src/app/app.module.ts [hunk 6]
```diff
@@ -98,7 +98,8 @@ import { ServiceDiscoveryComponent } from './service-discovery/service-discovery
     AuditService,
     SchedulesService,
     ServicesHealthService,
-    AssetSummaryService
+    AssetSummaryService,
+    StatusService
```
Comment (praveen-garg) [inline src/app/app.module.ts]: is this ConnectedServiceStatus

**Comment by mshariq-nerd** [comment]
@praveen-garg  PR fixed as per feedback. Please review now.

File: src/app/service-discovery/service-discovery.component.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+  ngAfterViewInit() {
+    this.isConnected = JSON.parse(localStorage.getItem('CONNECTED_SERVICE_STATE'));
+    this.connectedButtonId = localStorage.getItem('CONNECTED_SERVICE_ID');
+  }
+
+  setServiceDiscoveryURL() {
+    const protocolField = <HTMLSelectElement>document.getElementById('protocol');
+    const hostField = <HTMLInputElement>document.getElementById('host');
+    const servicePortField = <HTMLInputElement>document.getElementById('port');
+    const discovery_service_endpoint = protocolField.value + '://' + hostField.value + ':' + servicePortField.value + '/foglamp/discover';
```
Comment (praveen-garg) [inline src/app/service-discovery/service-discovery.component.ts]: discoveryServiceEndpoint

File: src/app/service-discovery/service-discovery.component.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    localStorage.setItem('DISCOVERY_SERVICE_URL', discovery_service_endpoint);
+    this.alertService.success('Discovery service url configured successfully.');
+  }
+
+  discoverService() {
+    let serviceRecord = [];
+    this.discoveryService.servicesDiscover()
```
Comment (praveen-garg) [inline src/app/service-discovery/service-discovery.component.ts]: `this.discoveryService.discover()` would be a better call; so def name change

File: src/app/service-discovery/service-discovery.component.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    this.isConnected = true;
+    localStorage.setItem('CONNECTED_SERVICE_STATE', JSON.stringify(this.isConnected));
+    localStorage.setItem('CONNECTED_SERVICE_ID', this.connectedButtonId);
+
+    let port = service[service.key].port
+    let address = service[service.key].addresses.length > 1 ? service[service.key].addresses[1] : service[service.key].addresses[0];
+
+    // TODO: Get protocol from service discovery
+    const service_endpoint = 'http://' + address + ':' + '8081/foglamp/';
+    const management_endpoint = 'http://' + address + ':' + port + '/foglamp/';
```
Comment (praveen-garg) [inline src/app/service-discovery/service-discovery.component.ts]: service_endpoint serviceEndpoint
management_endpoint managementEndpoint

File: src/app/service-discovery/service-discovery.component.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+    localStorage.setItem('SERVICE_URL', service_endpoint);
+    localStorage.setItem('MANAGEMENT_URL', management_endpoint);
+    location.reload();
+    location.href = '';
+    this.router.navigate([location.href]);
+  }
+
+  public closeMessage() {
+    console.log('hidden');
```
Comment (praveen-garg) [inline src/app/service-discovery/service-discovery.component.ts]: remove this log

File: src/app/services/discovery.service.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+@Injectable()
+export class DiscoveryService {
+
+  constructor(private http: Http) { }
+
+  /**
+   *  GET  | /foglamp/discover
+   */
+  servicesDiscover() {
+    const SERVICE_DISCOVER_URL = localStorage.getItem('DISCOVERY_SERVICE_URL');
```
Comment (praveen-garg) [inline src/app/services/discovery.service.ts]: SERVICE_DISCOVERY_URL and make sure it exists else log error etc

File: src/app/services/discovery.service.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+@Injectable()
+export class DiscoveryService {
+
+  constructor(private http: Http) { }
+
+  /**
+   *  GET  | /foglamp/discover
+   */
+  discover() {
+    const SERVICE_DISCOVER_URL = localStorage.getItem('DISCOVERY_SERVICE_URL');
```
Comment (praveen-garg) [inline src/app/services/discovery.service.ts]: make it local  ` const serviceDiscoveryUrl`

**Pull request merged**